### PR TITLE
Merge React 18 feature branch

### PR DESCRIPTION
**Pull request by JamesBurnside**
# What

Merge of base PRs:

- #2988
- #3005
- #3045

# Why

Merge short term solution for react 18, see https://github.com/Azure/communication-ui-library/discussions/2939 & https://github.com/Azure/communication-ui-library/issues/1900.

# How Tested

Selfhosting and bug bashes.

### CODE REVIEW

File: packages/storybook/stories/storybook.test.ts
Comment (JamesBurnside) [inline packages/storybook/stories/storybook.test.ts:48]: these are not working on react 18, they fail with a non-descript "I give up" error ☹️

File: packages/acs-ui-common/package.json
Comment (PorterNan) [inline packages/acs-ui-common/package.json:46]: Do we need to run react 16 compatibility test by any chance?
BTW, why not use ^ to get latest update for react 18

File: packages/communication-react/package.json
Comment (PorterNan) [inline packages/communication-react/package.json:68]: Just out of curiosity, why do we need to do this copy and paste?

File: packages/northstar-wrapper/src/index.ts
Comment (PorterNan) [inline packages/northstar-wrapper/src/index.ts:4]: Any concerns that we don't do import * ? is it for treeshaking

File: packages/northstar-wrapper/package.json
Comment (PorterNan) [inline packages/northstar-wrapper/package.json:6]: I guess we will generate both esm and cjs code in our npm package? How much bigger our npm packages will be in the end

File: packages/northstar-wrapper/package.json
Comment (PorterNan) [inline packages/northstar-wrapper/package.json:11]: If that is esm, will webpack bundle them into one file or several files like all the other packages?

File: packages/acs-ui-common/package.json
Comment (JamesBurnside) [reply packages/acs-ui-common/package.json:46]: This should be pinned to the same as react so this should be whatever react is. Really they should probably all be `~` or `^` but we don't have good repo-wide guidenence on what developers should choose. Should we set a standard now and update guidence in the wiki?

File: packages/communication-react/package.json
Comment (JamesBurnside) [reply packages/communication-react/package.json:68]: We need the webpack output from the build. Currently our meta package rebuilds everything itself and hence tsc doesn't use the webpack output. This was the simplest way I found to have the webpack bits in the final npm package (but open if you find a way to instead have tsc grab the webpack output instead of retranspiling the src code)

File: packages/northstar-wrapper/src/index.ts
Comment (JamesBurnside) [reply packages/northstar-wrapper/src/index.ts:4]: The question is why we would do `*` at all. We don't want to allow exports from `n*` so we want to keep this to a minimal set such that we don't allow developers to accidentally introduce new reliances on northstar

File: packages/northstar-wrapper/package.json
Comment (JamesBurnside) [reply packages/northstar-wrapper/package.json:11]: Still one file, but a file that has treeshaking support (treeshaking can still strip out unused exports in the same file)

File: packages/acs-ui-common/package.json
Comment (JamesBurnside) [reply packages/acs-ui-common/package.json:46]: And yeah we will need that, I have something cooking up on the side but haven't had time for it, will be similar to the typescript check
image

File: packages/storybook/stories/storybook.test.ts
Comment (emlynmac) [reply packages/storybook/stories/storybook.test.ts:48]: Maybe storybook 7 works?

**Review by emlynmac — COMMENTED** [review]
Just some questions, but looks good!

File: packages/react-composites/tests/browser/call/hermetic/VerticalGalleryResizing.test.ts
Comment (JamesBurnside) [reply packages/react-composites/tests/browser/call/hermetic/VerticalGalleryResizing.test.ts:90]: nope, it waits for the correct count and fails if the condition isn't met. Change here allows two things:
- the count is now waited for (so can be in subsequent render passes)
- the action will now screenshot on a failure

File: packages/react-composites/tests/browser/snapshots/beta/tests/browser/call/hermetic/ParticipantPane.test.ts-snapshots/PSTN-call-screen-with-dialpad-Mobile-Android-Landscape-linux.png
Comment (JamesBurnside) [reply packages/react-composites/tests/browser/snapshots/beta/tests/browser/call/hermetic/ParticipantPane.test.ts-snapshots/PSTN-call-screen-with-dialpad-Mobile-Android-Landscape-linux.png:1]: As far as I can tell its a react 18 optimization. Before with react 16 it took an extra render pass before that focus shows up and our tests took the screenshot before that was happening. With react 18 it handles bulk state changes better and passes focus more efficiently so they tests happen to take their screenshot after the component has taken focus. From running locally the focus is expected

File: packages/storybook/stories/storybook.test.ts
Comment (JamesBurnside) [reply packages/storybook/stories/storybook.test.ts:48]: Hopefully, although we actually were considering removing these tests as no one has reported finding a regression with them and they aren't fun to have to update or review - but that's a side conversation

File: packages/acs-ui-common/package.json
Comment (PorterNan) [reply packages/acs-ui-common/package.json:46]: > 

Yea, would be nice to always get latest bit of the current compatible version, cause it might be something unexpected happening there (very rare though) then we can catch it earlier than our customer (cause customers prefer those "^" for versions)

Comment (JamesBurnside) [reply packages/acs-ui-common/package.json:46]: I will keep this change for a separate PR, made a bug for it for tracking [Bug 3279379](https://skype.visualstudio.com/SPOOL/_workitems/edit/3279379): [Web] Repo should have guidence on using a pinned dep version vs `^` or `~`

File: packages/react-components/src/components/MessageStatusIndicator.tsx
Comment (dmceachernmsft) [inline packages/react-components/src/components/MessageStatusIndicator.tsx:126]: Small thing but did we do any announcer testing here? we have had bugs around this component in the past not announcing states

Comment (JamesBurnside) [reply packages/react-components/src/components/MessageStatusIndicator.tsx:126]: Just the testing done here: https://github.com/Azure/communication-ui-library/pull/3005. Confirmed the DOM shows the same behavior as before

File: packages/react-components/src/components/Announcer/AnnouncerContext.ts
Comment (JamesBurnside) [reply packages/react-components/src/components/Announcer/AnnouncerContext.ts:15]: I will leave this to a later discussion. This doesn't add a new provider, just moves it from an external package into our codebase

File: packages/northstar-wrapper/src/index.ts
Comment (JamesBurnside) [reply packages/northstar-wrapper/src/index.ts:4]: Nothing needs done it will automatically work - but actually wih it didn't. Someone _**should not**_ be adding to this list. We want to get further away from n* not be more reliant on it,

File: packages/communication-react/tsconfig.preprocess.json
Comment (PorterNan) [inline packages/communication-react/tsconfig.preprocess.json:14]: This is for preprocess version, do we preprocess northstar as well?

File: packages/communication-react/package.json
Comment (PorterNan) [reply packages/communication-react/package.json:68]: Yea.. within the internal packages, it will try to use src folder instead (and that's why we use copy-preprocess), btw, I guess doing copy-northstar in the end means we want it to be packaged and consumed by customer, but our own build process doesn't need it?

File: packages/react-composites/src/composites/CallComposite/MockCallAdapter.ts
Comment (dmceachernmsft) [inline packages/react-composites/src/composites/CallComposite/MockCallAdapter.ts:101]: why do we need an implementation here now? is there one needed for createView as well?

File: packages/northstar-wrapper/package.json
Comment (PorterNan) [reply packages/northstar-wrapper/package.json:6]: Yea... I agree with production build should always use esm, but still there would be some app living in cjs world for some legacy reasons, so we might still want to support that well

For npm package, that is just a question out of curiosity, I did receive a feedback from webchat team asking about the  npm package size, but it shouldn't matter too much in the end

File: packages/communication-react/package.json
Comment (JamesBurnside) [reply packages/communication-react/package.json:68]: Yeah that's correct, it does mean we have gaps where parts of our build system is using the webpack output and other parts just directly grabs the n* code from the northstar-wrapper/node_modules :/

File: packages/react-composites/tests/app/lib/HiddenChatComposites.tsx
Comment (dmceachernmsft) [inline packages/react-composites/tests/app/lib/HiddenChatComposites.tsx:15]: maybe dumb question but what is the reason for the hidden composites again? just to send messages in the tests? (wont block this PR on this but would love to know why we have these)

File: packages/communication-react/tsconfig.preprocess.json
Comment (JamesBurnside) [reply packages/communication-react/tsconfig.preprocess.json:14]: There's no conditional compiles for n*, and given we should not be introducing new usage of this library there shouldn't be in future so stable vs beta didn't need supported in this package

File: packages/react-composites/tests/browser/call/hermetic/HorizontalGalleryResizing.test.ts
Comment (dmceachernmsft) [inline packages/react-composites/tests/browser/call/hermetic/HorizontalGalleryResizing.test.ts:90]: love this refactoring

File: packages/communication-react/tsconfig.preprocess.json
Comment (JamesBurnside) [reply packages/communication-react/tsconfig.preprocess.json:14]: (so I guess this line isn't needed? I just copy-pasted anywhere `@internal` packages were referenced)

File: packages/northstar-wrapper/src/index.ts
Comment (PorterNan) [reply packages/northstar-wrapper/src/index.ts:4]: Are we saying we developers could import NorthStar component from our package? I guess we don't expose them as a public api right?

File: packages/react-components/src/components/Announcer/Announcer.tsx
Comment (PorterNan) [inline packages/react-components/src/components/Announcer/Announcer.tsx:1]: The reason why we have these files is because it's not exportable from northstar, I guess?

File: packages/northstar-wrapper/package.json
Comment (JamesBurnside) [reply packages/northstar-wrapper/package.json:6]: True, and we do support cjs, we just don't optimize any bundles. However anyone using this in cjs will actually receive a benefit because they don't have to rollup the entire n* library, the small pieces needed are now directly bundled by us so smaller package overall (unless of course they use both us and n* in their app and are bundling using cjs)

File: packages/react-components/src/components/VideoGallery/utils/videoGalleryLayoutUtils.test.tsx
Comment (PorterNan) [inline packages/react-components/src/components/VideoGallery/utils/videoGalleryLayoutUtils.test.tsx:326]: This is the new format from new react test package?

File: packages/northstar-wrapper/src/index.ts
Comment (JamesBurnside) [reply packages/northstar-wrapper/src/index.ts:4]: nope, just internal developers to this repo, not npm package users

File: packages/react-components/src/components/VideoGallery/utils/videoGalleryLayoutUtils.test.tsx
Comment (PorterNan) [reply packages/react-components/src/components/VideoGallery/utils/videoGalleryLayoutUtils.test.tsx:326]: Sry the render one

Comment (PorterNan) [reply packages/react-components/src/components/VideoGallery/utils/videoGalleryLayoutUtils.test.tsx:326]: Ohh, you mentioned react 18 change its way to render components

File: packages/react-components/src/components/Announcer/Announcer.tsx
Comment (JamesBurnside) [reply packages/react-components/src/components/Announcer/Announcer.tsx:1]: See here: https://github.com/Azure/communication-ui-library/pull/3005. The component we used has react 16 as peer dependency but it is a tiny package so we are just taking the code from it directly

File: packages/react-components/src/components/VideoGallery/utils/videoGalleryLayoutUtils.test.tsx
Comment (JamesBurnside) [reply packages/react-components/src/components/VideoGallery/utils/videoGalleryLayoutUtils.test.tsx:326]: yeah

File: packages/react-composites/tests/app/chat/index.tsx
Comment (PorterNan) [inline packages/react-composites/tests/app/chat/index.tsx:19]: Yay, with this code piece, I guess we are good to use newest CRA to create projects!

File: packages/react-components/src/components/VideoGallery/utils/videoGalleryLayoutUtils.test.tsx
Comment (JamesBurnside) [reply packages/react-components/src/components/VideoGallery/utils/videoGalleryLayoutUtils.test.tsx:326]: Yeah newer syntax with the newer testing library. the render blocks don't need to be in act blocks

File: packages/react-composites/tests/app/chat/index.tsx
Comment (JamesBurnside) [reply packages/react-composites/tests/app/chat/index.tsx:19]: We are!

File: packages/react-composites/tests/browser/snapshots/beta/tests/browser/call/hermetic/ParticipantPane.test.ts-snapshots/PSTN-call-screen-with-dialpad-Mobile-Android-Landscape-linux.png
Comment (dmceachernmsft) [inline packages/react-composites/tests/browser/snapshots/beta/tests/browser/call/hermetic/ParticipantPane.test.ts-snapshots/PSTN-call-screen-with-dialpad-Mobile-Android-Landscape-linux.png:1]: Did we always have this CSS when the box has focus? I thought the underline just turned blue

File: packages/react-composites/src/composites/CallComposite/MockCallAdapter.ts
Comment (JamesBurnside) [reply packages/react-composites/src/composites/CallComposite/MockCallAdapter.ts:101]: One of the tests called it as the tile logic now completes its unmount before the test finishes

File: packages/react-composites/tests/app/lib/HiddenChatComposites.tsx
Comment (JamesBurnside) [reply packages/react-composites/tests/app/lib/HiddenChatComposites.tsx:15]: Yeah just to send messages and read receipts. They don't really have to be hidden tbh, and I wonder if we could replace them with just adapter uses and fire off read events and messages and such through the adapter interface

File: packages/react-composites/tests/browser/call/hermetic/VerticalGalleryResizing.test.ts
Comment (PorterNan) [reply packages/react-composites/tests/browser/call/hermetic/VerticalGalleryResizing.test.ts:90]: So previous one will be waiting for dataUiId(IDS.verticalGalleryVideoTile), if that is showing up, it will start counting the number

But the current behavior would be wait until it is 2 count, and failed it that is not (and take screenshot I guess?)

File: packages/react-composites/tests/browser/snapshots/beta/tests/browser/call/hermetic/ParticipantPane.test.ts-snapshots/PSTN-call-screen-with-dialpad-Mobile-Android-Landscape-linux.png
Comment (JamesBurnside) [reply packages/react-composites/tests/browser/snapshots/beta/tests/browser/call/hermetic/ParticipantPane.test.ts-snapshots/PSTN-call-screen-with-dialpad-Mobile-Android-Landscape-linux.png:1]: I believe you already asked this one 😉 https://github.com/Azure/communication-ui-library/pull/3045#issuecomment-1546025173

File: packages/react-composites/tests/browser/call/hermetic/VerticalGalleryResizing.test.ts
Comment (PorterNan) [reply packages/react-composites/tests/browser/call/hermetic/VerticalGalleryResizing.test.ts:90]: I am good with this one, this might miss some "delay render"(like first second it renders one element, but later render 2 elements which meet the condition) issues, but also allow use to get more stable UI snapshot

It could have either make our pipeline faster or slower, would like to see the final number when we have this in main

Comment (JamesBurnside) [reply packages/react-composites/tests/browser/call/hermetic/VerticalGalleryResizing.test.ts:90]: I believe the previous one actually did no waiting, it just immediately checked what was rendered. We now instead wait for the condition to become true because the resize observer kicks in after mounting so the tiles we show changes across a couple render passes

File: packages/react-composites/tests/browser/callwithchat/hermetic/Chat.test.ts
Comment (PorterNan) [inline packages/react-composites/tests/browser/callwithchat/hermetic/Chat.test.ts:49]: Looks like we are typing 'I agree' for twice and I guess this is changing the test behaivor?

File: packages/react-composites/tests/browser/common/hermeticChatTestHelpers.ts
Comment (PorterNan) [inline packages/react-composites/tests/browser/common/hermeticChatTestHelpers.ts:20]: Maybe we need a hidden element in html to notify that "I have successfully sent the read receipt at what time point"

File: packages/react-composites/tests/browser/snapshots/beta/tests/browser/call/hermetic/ParticipantPane.test.ts-snapshots/PSTN-call-screen-with-dialpad-Mobile-Android-Landscape-linux.png
Comment (PorterNan) [inline packages/react-composites/tests/browser/snapshots/beta/tests/browser/call/hermetic/ParticipantPane.test.ts-snapshots/PSTN-call-screen-with-dialpad-Mobile-Android-Landscape-linux.png:1]: Noticed there is a focus change here, is it by designed?

File: packages/react-composites/tests/browser/callwithchat/hermetic/Chat.test.ts
Comment (JamesBurnside) [reply packages/react-composites/tests/browser/callwithchat/hermetic/Chat.test.ts:49]: The logic for how the hidden composite is hidden has changed. We now no longer render it entirely whereas before we just marked visibility as hidden. This is because react is much better now and hooks like triggering read receipts get triggered when the composite is hidden now. Because we unmount anything typed and not sent it lost, so I've updated the functions to by type* and typeAndSend*

File: packages/react-composites/tests/browser/common/hermeticChatTestHelpers.ts
Comment (JamesBurnside) [reply packages/react-composites/tests/browser/common/hermeticChatTestHelpers.ts:20]: Yeah :/ Would be nice to use the adapter and not composites for these actually and use adapter interface to send the read receipt

File: packages/react-composites/tests/browser/snapshots/beta/tests/browser/call/hermetic/ParticipantPane.test.ts-snapshots/PSTN-call-screen-with-dialpad-Mobile-Android-Landscape-linux.png
Comment (JamesBurnside) [reply packages/react-composites/tests/browser/snapshots/beta/tests/browser/call/hermetic/ParticipantPane.test.ts-snapshots/PSTN-call-screen-with-dialpad-Mobile-Android-Landscape-linux.png:1]: Yep see here: https://github.com/Azure/communication-ui-library/pull/3045#issuecomment-1546025173

File: packages/northstar-wrapper/package.json
Comment (PorterNan) [reply packages/northstar-wrapper/package.json:11]: Lets open a work item to validate that later including find a solution for treeshaking localization part

File: packages/react-composites/tests/browser/callwithchat/hermetic/Chat.test.ts
Comment (PorterNan) [reply packages/react-composites/tests/browser/callwithchat/hermetic/Chat.test.ts:49]: Could be a work item not to use hidden composite but adapter/clients

File: packages/react-composites/src/composites/CallComposite/MockCallAdapter.ts
Comment (dmceachernmsft) [reply packages/react-composites/src/composites/CallComposite/MockCallAdapter.ts:101]: ahhh okay that makes sense

File: packages/react-composites/tests/app/lib/HiddenChatComposites.tsx
Comment (dmceachernmsft) [reply packages/react-composites/tests/app/lib/HiddenChatComposites.tsx:15]: yeah could be a interesting refactoring of the tests. I don't think its anything worth blocking this over though since are already using this.

File: packages/react-composites/tests/browser/snapshots/beta/tests/browser/call/hermetic/ParticipantPane.test.ts-snapshots/PSTN-call-screen-with-dialpad-Mobile-Android-Landscape-linux.png
Comment (dmceachernmsft) [reply packages/react-composites/tests/browser/snapshots/beta/tests/browser/call/hermetic/ParticipantPane.test.ts-snapshots/PSTN-call-screen-with-dialpad-Mobile-Android-Landscape-linux.png:1]: I did and totally missed it! that makes sense to me. I am also just remembering the CSS as it turns out!

**Review by PorterNan — APPROVED** [review]

**Review by dmceachernmsft — APPROVED** [review]

File: samples/Chat/tsconfig.json
Comment (JamesBurnside) [reply samples/Chat/tsconfig.json:12]: Good question! We avoid imports in the code to make sure we only use publicly available exports so that when we migrate the code to the samples nothing breaks. But internally it needs to reference the packlets for quicker building (doesn't have to first build the downstream packlets) and for hot reloading

**Review by emlynmac — APPROVED** [review]

**Pull request merged**
